accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,113 @@
+
+### Runtime/libraries already have some feature switches
+Globalization has a feature where it can always use only invariant culture and nothing else. For reference: [design doc](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md) and the [base of the implementation](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs). This has SDK integration to set it up in `.runtimeconfig.json` but it doesn't have linker integration.
+
+There's also a discussion to setup the build of runtime/libraries to automatically trim based on target configuration - [here](https://github.com/dotnet/runtime/issues/31785). The intent of this change is not feature switches, but it's in the same area.
+
+### Runtime/libraries already have compatibility switches
+For example `CoreLib` already has several switches like `Switch.System.Globalization.FormatJapaneseFirstYearAsNumber`. These are typically small "tweaks" to the behavior of the runtime/libraries. They are internally exposed as static properties on [`LocalAppContextSwitches`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.cs) with the value coming from `.runtimeconfig.json` via [`AppContext.TryGetSwitch`](https://docs.microsoft.com/en-us/dotnet/api/system.appcontext.trygetswitch?view=netcore-3.1).
+
+
+## Proposal
+The purpose of the proposed changes is to introduce patterns and functionality to support them to make introduction of new feature switches easier and consistent.
+
+### Unified pattern for defining switches and their properties
+Each feature switch should have a read-only static property defined and all the decisions about the feature's availability should be based on a simple condition using such property. Depending on the impact of the feature switch the property may or may not be a public API.
+
+Feature switches for large areas of functionality which may span multiple libraries should be exposed as public APIs to make it easy and consistent. For example if there would be a feature switch to disable support for HTTP2 there should be a public API property indicating the desired behavior. This is because it's likely that even libraries outside of core libraries may need to change their behavior based on the property. Having one property in this case would enable linker to trim branches across multiple libraries in a consistent way.
+
+Feature switches for localized functionality which is not likely to span multiple libraries can use only internal properties.
+
+*Note: `CoreLib` for example already has `LocalAppContextSwitches` which does basically exactly this. We could simply extend it to usage of feature switches not just compatibility switches.*
```

Comment:
From a quick glance, it doesn't look like corelib's current structure for LocalAppContextSwitches would make it amenable to tiering optimizations that would let the JIT trim away "dead" code resulting from a switch being set.  It'd be nice if the recommended structure for these were amenable to that, e.g. that a static readonly bool would end up being the thing conditionals were based on.  This won't matter in many cases, but it could for checks that show up on hot paths.  GlobalizationMode.Invariant is a good example of this: it is accessed on hot paths, and it's just a property that returns a static readonly bool field.

Reply:
OK - I will change that - should I copy the GlobalizationMode.Invariant - or is there some typical pattern which JIT can figure out?

Reply:
`GlobalizationMode.Invariant` is a fine example. Property backed by a readonly bool field is the key pattern.

The LocalAppContextSwitches pattern has ability to [disable caching](https://github.com/dotnet/runtime/blob/0a2f0e6da98a784601b37d19417b930a33b2c478/src/libraries/System.Private.CoreLib/src/System/LocalAppContextSwitches.Common.cs#L33). This is used in [a few tests](https://github.com/dotnet/runtime/search?q=DisableCaching&unscoped_q=DisableCaching). The ability to disable caching is what makes the pattern unfriendly to linking. I do not see a problem with giving up the disable caching hook and implement the tests in a different way.